linera-witty/tests/reentrancy.rs

```diff
@@ -0,0 +1,485 @@
+    {
+        ImportedGetters::new(caller).get_u32()
+    }
+
+    fn get_s64<Caller>(caller: Caller) -> Result<i64, RuntimeError>
+    where
+        Caller: InstanceForImportedGetters,
+        <Caller::Runtime as Runtime>::Memory: RuntimeMemory<Caller>,
+    {
+        ImportedGetters::new(caller).get_s64()
+    }
+
+    fn get_u64<Caller>(caller: Caller) -> Result<u64, RuntimeError>
+    where
+        Caller: InstanceForImportedGetters,
+        <Caller::Runtime as Runtime>::Memory: RuntimeMemory<Caller>,
+    {
+        ImportedGetters::new(caller).get_u64()
+    }
+
+    fn get_float32<Caller>(caller: Caller) -> Result<f32, RuntimeError>
+    where
+        Caller: InstanceForImportedGetters,
+        <Caller::Runtime as Runtime>::Memory: RuntimeMemory<Caller>,
+    {
+        ImportedGetters::new(caller).get_float32()
+    }
+
+    fn get_float64<Caller>(caller: Caller) -> Result<f64, RuntimeError>
+    where
+        Caller: InstanceForImportedGetters,
+        <Caller::Runtime as Runtime>::Memory: RuntimeMemory<Caller>,
+    {
+        ImportedGetters::new(caller).get_float64()
+    }
+}
+
+/// Test reentrant functions with return values.
+///
+/// The host functions are called from the guest, and they return values obtained by calling back
+/// the guest through functions with the same names.
+#[test_case(MockInstanceFactory::default(); "with a mock instance")]
+#[cfg_attr(feature = "wasmer", test_case(WasmerInstanceFactory; "with Wasmer"))]
+#[cfg_attr(feature = "wasmtime", test_case(WasmtimeInstanceFactory; "with Wasmtime"))]
+fn getters<InstanceFactory>(mut factory: InstanceFactory)
+where
+    InstanceFactory: TestInstanceFactory,
+    InstanceFactory::Instance: InstanceForEntrypoint,
+    <<InstanceFactory::Instance as Instance>::Runtime as Runtime>::Memory:
+        RuntimeMemory<InstanceFactory::Instance>,
+    ExportedGetters: ExportTo<InstanceFactory::Builder>,
+{
+    let instance = factory.load_test_module("reentrancy", "getters", |instance| {
+        ExportedGetters::export_to(instance)
+            .expect("Failed to export reentrant getters WIT interface")
+    });
+
+    Entrypoint::new(instance)
+        .entrypoint()
+        .expect("Failed to call guest's `entrypoint` function");
+}
+
+/// An interface to import functions with parameters.
+#[wit_import(package = "witty-macros:test-modules", interface = "setters")]
+trait ImportedSetters {
+    fn set_bool(value: bool);
+    fn set_s8(value: i8);
+    fn set_u8(value: u8);
+    fn set_s16(value: i16);
+    fn set_u16(value: u16);
+    fn set_s32(value: i32);
+    fn set_u32(value: u32);
+    fn set_s64(value: i64);
+    fn set_u64(value: u64);
+    fn set_float32(value: f32);
+    fn set_float64(value: f64);
+}
+
+/// Type to export reentrant functions with parameters.
+pub struct ExportedSetters;
+
+#[wit_export(package = "witty-macros:test-modules", interface = "setters")]
+impl ExportedSetters {
```

Comment:
Would it make sense to support this more concise variant?
```
pub struct ExportedSetters<Caller>(std::marker::PhantomData<Caller>));

impl Default for ExportedSetters<Caller> { ... }

#[wit_export(package = "witty-macros:test-modules", interface = "setters")]
impl<Caller> ExportedSetters<Caller>
where
    Caller: InstanceForImportedSetters,
    <Caller::Runtime as Runtime>::Memory: RuntimeMemory<Caller>,
{
    fn set_bool(caller: Caller, value: bool) -> Result<(), RuntimeError> { ... }

    // ..
}
```

Reply:
That would be better, thanks! I'll look into it :+1:

Reply:
Well, I started doing it, but it's a bit more complex than I thought :sweat_smile: . I opened #963 to finish it later.